addons/l10n_be_edi/models/mail_template.py

```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+
+from odoo import models
+
+from lxml import etree
+import base64
+from xml.sax.saxutils import escape, quoteattr
+
+
+class MailTemplate(models.Model):
+    _inherit = "mail.template"
+
+    def _l10n_be_edi_inject_pdf_into_xml(self, pdf_name, pdf, xml):
+        '''Adds the pdf report in the e-fff XML as base64 string.
+        '''
+        to_inject = '''
+            <cac:AdditionalDocumentReference xmlns="urn:oasis:names:specification:ubl:schema:xsd:Invoice-2" xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2" xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2">
+                <cbc:ID>%s</cbc:ID>
+                <cac:Attachment>
+                    <cbc:EmbeddedDocumentBinaryObject mimeCode="application/pdf" filename=%s>
+                        %s
+                    </cbc:EmbeddedDocumentBinaryObject>
+                </cac:Attachment>
+            </cac:AdditionalDocumentReference>
+        ''' % (escape(pdf_name), quoteattr(pdf_name), pdf.decode())
+
+        old_xml = base64.b64decode(xml, validate=True)
+        tree = etree.fromstring(old_xml)
+        contentnav = tree.xpath("//*[local-name()='DocumentCurrencyCode']")[0]
+        contentnav.addnext(etree.fromstring(to_inject))
+        return etree.tostring(tree, pretty_print=True)
+
+    def _l10n_be_find_pdf(self, attachments_dict):
+        ''' Find the first attachment that is a pdf based on the name's extension and returns the name.
+        '''
+        for name in attachments_dict.keys():
+            if name.endswith('.pdf'):
+                return name
+        return False
+
+    def generate_email(self, res_ids, fields):
+        ''' OVERRIDE For the Belgian EDI (e-fff), it is required to add the pdf inside the xml.
```

Comment:
Wait, we have a generic mail.template system, and a generic report template system, and they can interoperate correctly. Yet here you're tinkering with the generation and plugging your stuff programmatically. That kind of defeats the whole purpose of having generic systems.
It also means that the report is kind of broken if you use it without that "mail template magic".
 
Imagine if every model decided to overrides the mail.template rendering to plug in their own reports and post-process attachments like this. It would quickly become a complete mess of Python overrides with some of them having side-effects on what the others are doing, and all hell breaks loose.

Also , arbitrarily taking the first PDF attached and embedding it blindly is quite brittle. What if the admin decides to add some "Promo.PDF" to the invoice notification mail template? Are you going to embed this doc in the XML?


Really, if one of the reports embeds the rendering of another report, then so be it, it's going to have to call the other while preparing its data, no? Besides, the invoice PDF should be cached as an attachment once rendered, so you might even save the double-rendering in many cases.

By doing this, all reports will behave as expected even when used outside of a `mail.template` rendering, and you will remove many sources of problems down the road.

PS: I wanted to check the specs for some rationale that could explain this design, but could not seem to find the task ID in the PR ;-)

Reply:
@odony The xml is not a report, rather it's generated by the account_edi system. The thing is, when it's generated, the pdf is not yet generated. Since this is very specific to belgium, we try to avoid modifying both the account_edi flow and the account_invoice_send flow (pdf is generated when the user clicks on "send and print" for the first time, this is taken care of by the account_invoice_send wizard). So the idea was to generate the xml in the account_edi flow but hiding it from the user and when we generate the invoice's report, we inject it in the xml file before generating the email.

The way to get the pdf is a bit fishy too, I agree, but since it's called in account_invoice_send wizard I think it's ok. The thing is, the override is done in generate_mail because account_invoice_send has a compose inherit of it. I would prefer doing it in account_invoice_send, but I don't think that's possible.

As for caching, I prefer to keep the data as raw as possible, because both the xml and the pdf can be edited in case of a reset to draft and that would result in inconsistencies or even more 'bidouille'.

I'm still looking to see if I can improve stuff based on your comment.

PS : I added the ID of the task, sorry for that ;)